README.md
# Parallel Playback

This is a tool for synchronizing media playback (video or audio) on
multiple remote clients.  *Viewer* clients load media files from an
arbitrary external source, and then start and stop playback according
to signals from a *host* client.

This takes the form of an HTTP server which serves single-page
javascript clients for the viewers and host.  The clients in turn
stream media from any direct HTTP link, in any format that can be
played by a browser (`.webm`, `.opus`, `.mp4`, `.mp3`, etc. depending
on browser).

## Getting started

Building the `parallel-playback` server requires the [Rust
toolchain][install-rust].  From the root of this repository, run:

    $ cargo run -- 8000 ./test-config.json ./clients

You may now find the host client at
`https://localhost:8000/host/?key=zcmdzdx8` and the viewer client at
`https://localhost:8000/join/?key=pf9fflsk`.  Please note the requirement for
`https` ([details](#https-and-wss)).

The javascript clients (in `./clients`) are simple HTML pages and
scripts; they do not require any sort of compiling or processing.

### Command arguments and configuration

The `parallel-playback` server program takes three arguments: the port
to connect on, the configuration file path, and the path to static
client files.

The configuration file gives a list of distinct "parties", with keys
to connect as host (host) and as viewer (join).

    {
        "parties": [
            { "host": "zcmdzdx8", "join": "pf9fflsk" }
            { "host": "7g5vpd9r", "join": "8qq8fs2x" }
        ]
    }

The host clients of one party only interact with that party's
join/viewer clients.  Multiple host clients for a single party can be
open; Play and Pause signals sent from any host will have the same
effect.

Depending on your use case, you may want to share both the host and
join keys with party participants, or just the join key.

## Wait, so what is this thing?

The main purpose of this tool is to support a "watch-together" type
remote party, in which remote participants watch video or listen to
music while talking to each other over a video-call.  In particular,
its design solves three key problems:

1. The media comes from an external source, so the host participant's
   video-call stream is not squashed.  Media quality is also not
   limited by the video-call quality.
2. Because the media stream is distinct from the video call, viewers
   can control its volume separately.
3. "Play" signals schedule a start time several seconds in the future,
   which viewer clients match to their local clock.  This means that
   start times will not be skewed by large ping times between the host
   and viewers.

Another use-case is to play background music and effects for a
tabletop-style roleplaying game session.

## What to watch

Direct links to video and audio are not exactly common on the web, so
if you have a particular movie or soundtrack in mind to play using
this tool, you may need to set up an HTTP host for it yourself.

But if you want to use existing hosted content, there are options:

- If you're looking for music and effects for a game session, check
  out [OpenGameArt.org][oga-music].
- Podcasts must always directly host audio files, so if a podcast
  listening-party is your thing, your options are endless!  For
  example, use any download link on the [Libre Lounge archive
  page][lla].

## Troubleshooting

### HTTPS and WSS

This tool is intended to serve pages through HTTPS and connect to
clients using WSS; if you can't access `localhost` using HTTPS, you
will need to edit the `wss://` websocket links to `ws://` in the
[host](./clients/host/index.js) and [join](./clients/host/index.js)
clients for local testing (yuck, I know).  [This guide][localcert]
from Let's Encrypt has some suggestions for fixing this problem more
cleanly for your dev environment.

## License

This program is free software: you can redistribute it and/or modify
it under the terms of the [GNU General Public License][gpl3] as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the [GNU
General Public License][gpl3] for more details.

*Note: If you publish a modified version, you should edit the footers
of the [host](./clients/host/index.html) and
[join](./clients/join/index.html) clients to link to the repository of
your modified source code.*


[install-rust]: https://www.rust-lang.org/tools/install
[oga-music]: https://opengameart.org/art-search-advanced?keys=&field_art_type_tid%5B%5D=12&sort_by=count&sort_order=DESC
[lla]: https://librelounge.org/archive/
[localcert]: https://letsencrypt.org/docs/certificates-for-localhost/
[gpl3]: https://www.gnu.org/licenses/gpl-3.0.html


CHANGELOG.md
## v0.1.1 (2021-02-01)

Allow a WebVTT text track to be specified alongside media to act as
captions or subtitles.

## v0.1.0 (2021-01-30)

Initial release.


src/main.rs
mod comms;
use comms::State;
mod config;
use config::Config;
mod timing;

use std::env;
use std::path::Path;
use std::str::FromStr;

use warp::Filter;

#[tokio::main]
async fn main() {

    let args = env::args().collect::<Vec<String>>();
    if args.len() != 4 {
        panic!("Requires 3 arguments, got {}.", args.len() - 1);
    }
    let port = args[1].parse::<u16>().expect("Couldn't parse port.");
    let c = match Config::load(Path::new(&args[2])) {
        Ok(c) => c,
        Err(e) => panic!(e),
    };
    let static_dir = Path::new(&args[3]);
    let state = State::new(c.parties);
    let state2 = state.clone();

    let state_f = warp::any().map(move || state2.clone());

    let msg = warp::path("msg")
        .and(warp::ws())
        .and(state_f)
        .map(|ws: warp::ws::Ws, state| {
            ws.on_upgrade(move |socket|
                          comms::handle_connection(state,socket))
        });

    let common = warp::path("common")
        .and(warp::fs::dir(static_dir.join("common")));
    let join = warp::path("join")
        .and(warp::fs::dir(static_dir.join("join")));
    let host = warp::path("host")
        .and(warp::fs::dir(static_dir.join("host")));

    let routes = common.or(join).or(host).or(msg);

    // Listen on all ipv4 and ipv6 interfaces
    let addr = std::net::IpAddr::from_str("::0").unwrap();

    warp::serve(routes).run((addr, port)).await;
}


src/comms.rs
mod message;
use message::{IntroMsg,HostMsg,JoinerMsg};

use crate::config::Party;
use crate::timing;

use std::{
    collections::HashMap,
    sync::Arc,
};

use futures::{FutureExt, StreamExt};
use tokio::sync::{mpsc, RwLock};
use tokio_stream::wrappers::UnboundedReceiverStream;
use warp::ws::{Message, WebSocket};

type Tx = mpsc::UnboundedSender<Result<Message, warp::Error>>;

#[derive(Clone,Copy,PartialEq,Eq,PartialOrd,Ord,Debug)]
enum Role {
    Host(usize),
    Joiner(usize),
}

pub struct State {
    next_id: usize,
    conns: HashMap<usize, Tx>,
    parties: Vec<Party>,
    p_map: HashMap<usize,Role>,
}

pub type StateH = Arc<RwLock<State>>;

impl State {
    pub fn new(parties: Vec<Party>) -> StateH {
        StateH::new(RwLock::new(
            State{
                next_id: 0,
                conns: HashMap::new(),
                parties,
                p_map: HashMap::new(),
            }
        ))
    }
    pub fn insert(&mut self, tx: Tx) -> usize {
        let id = self.next_id;
        self.next_id += 1;
        self.conns.insert(id, tx);
        id
    }
    /// Upgrade connection to host of party.  Ok(()) on success,
    /// Err(()) when key matches no party's host-key.
    pub fn host(&mut self, id: usize, key: String) -> Result<usize,()> {
        for (p,ks) in self.parties.iter().enumerate() {
            if key == ks.host {
                self.p_map.insert(id, Role::Host(p));
                return Ok(p)
            }
        }
        Err(())
    }
    /// Upgrade connection to joiner of party.  Ok(()) on success,
    /// Err(()) when key matches no party's join-key.
    pub fn join(&mut self, id: usize, key: String) -> Result<usize,()> {
        for (p,ks) in self.parties.iter().enumerate() {
            if key == ks.join {
                self.p_map.insert(id, Role::Joiner(p));
                return Ok(p)
            }
        }
        Err(())
    }
    /// Get role (if any) of connection.
    fn role(&self, id: usize) -> Option<Role> {
        self.p_map.get(&id).map(|r| r.clone())
    }
    /// Get Vec of channels for joiners of given party_id
    pub fn pdests(&self, party_id: usize) -> Vec<Tx> {
        let mut out = Vec::new();
        for c in self.conns.iter() {
            match self.p_map.get(&c.0) {
                Some(Role::Joiner(p)) if party_id == *p =>
                    out.push(c.1.clone()),
                _ => {},
            }
        }
        out
    }
    pub fn remove(&mut self, id: usize) {
        self.conns.remove(&id);
        self.p_map.remove(&id);
    }
}

pub async fn broadcast(state: &StateH, msg: &str, party_id: usize) {
    let dests = &state.read().await.pdests(party_id);

    for tx in dests {
        match tx.send(Ok(Message::text(msg.to_string()))) {
            Ok(_) => {},
            Err(_) => {}, // tx is disconnected, no need to send
        }
    }
}

pub async fn handle_connection(
    state: StateH,
    ws: WebSocket,
) {
    let (ws_tx, mut ws_rx) = ws.split();
    let (tx,rx) = mpsc::unbounded_channel();
    let rx = UnboundedReceiverStream::new(rx);
    tokio::task::spawn(rx.forward(ws_tx).map(|result| {
        match result {
            Ok(_) => {},
            Err(e) => println!("Websocket send error: {}", e),
        }
    }));
    let conn_id = state.write().await.insert(tx);
    println!("New conn {}.", conn_id);

    while let Some(result) = ws_rx.next().await {
        let msg = match result {
            Ok(m) => match m.to_str() {
                Ok(s) => s.to_string(),
                Err(e) => {
                    println!("Msg not text? {:?}", e);
                    break;
                },
            },
            Err(e) => {
                println!("Websocket error(id={}): {}", conn_id, e);
                break;
            },
        };
        if msg != "ping" {
            let r = state.read().await.role(conn_id);
            match r {
                Some(Role::Host(p)) => match serde_json::from_str(&msg) {
                    Ok(HostMsg::Play(url,opt_vtt,seekstr,offset_secs)) =>
                       match timing::from_seekstr(&seekstr) {
                           Ok(seek_secs) => {
                               let m = serde_json::to_string(
                                   &JoinerMsg::Play(
                                       url,
                                       opt_vtt,
                                       seek_secs,
                                       timing::get_playtime(offset_secs)
                                   )
                               ).unwrap();
                               broadcast(&state, &m, p).await;
                           },
                           Err(e) => println!("Bad seekstr: {}", e),
                       },
                    Ok(HostMsg::Pause) => {
                        let m = serde_json::to_string(&JoinerMsg::Pause).unwrap();
                        broadcast(&state, &m, p).await;
                    },
                    Err(e) => println!("Could not decode as HostMsg. {}", e),
                },
                Some(Role::Joiner(_p)) => {
                    // Joiners don't send messages
                },
                None => match serde_json::from_str(&msg) {
                    Ok(IntroMsg::Host(key)) =>
                        match state.write().await.host(conn_id, key.clone()) {
                            Ok(p) => println!("{} is hosting {}.", conn_id, p),
                            Err(()) => println!("No match for key {}.", key),
                        },
                    Ok(IntroMsg::Join(key)) =>
                        match state.write().await.join(conn_id, key.clone()) {
                            Ok(p) => println!("{} is joining {}.", conn_id, p),
                            Err(()) => println!("No match for key {}.", key),
                        },
                    Err(e) => println!("Coud not decode as IntroMsg. {}", e),
                },
            }
        }
    }

    state.write().await.remove(conn_id);
    println!("Disconnect: {}", conn_id);
}


src/timing.rs
use std::num::ParseIntError;
use std::time::{SystemTime, UNIX_EPOCH};

/// Get a time to start playing the media, in milliseconds, which will
/// be now + the given number of seconds.
pub fn get_playtime(offset_secs: usize) -> u128 {
    let now = SystemTime::now();
    let now_millis = now
        .duration_since(UNIX_EPOCH)
        .expect("We've gone back in time?")
        .as_millis();
    now_millis + (offset_secs as u128 * 1000)
}

/// Decode hour:min:sec seek string into number of seconds.
pub fn from_seekstr(s: &str) -> Result<usize, ParseIntError> {
    let mut parts: Vec<&str> = s.split(":").collect();
    let hours = if parts.len() >= 3 {
        parts.remove(0).parse()?
    } else {
        0
    };
    let mins = if parts.len() >= 2 {
        parts.remove(0).parse()?
    } else {
        0
    };
    let secs = if parts.len() >= 1 {
        parts.remove(0).parse()?
    } else {
        0
    };
    Ok(hours * 60 * 60 + mins * 60 + secs)
}


src/comms/message.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub enum IntroMsg {
    Host(String),
    Join(String),
}

#[derive(Serialize, Deserialize, Debug)]
pub enum HostMsg {
    /// Play(url, vtt, seek_pos ("min:sec"), start_offset (seconds))
    Play(String, Option<String>, String, usize),
    Pause,
}

#[derive(Serialize, Deserialize, Debug)]
pub enum JoinerMsg {
    /// Play(url, vtt, seek_pos (seconds), start_time (milliseconds unix))
    Play(String, Option<String>, usize, u128),
    Pause,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn intromsg_rep() {
        assert_eq!(
            serde_json::to_string(&IntroMsg::Host("asdf".to_string())).unwrap(),
            "{\"Host\":\"asdf\"}",
        );
        assert_eq!(
            serde_json::to_string(&IntroMsg::Join("asdf".to_string())).unwrap(),
            "{\"Join\":\"asdf\"}",
        );
    }

    #[test]
    fn hostmsg_rep() {
        assert_eq!(
            serde_json::to_string(&HostMsg::Play("http://asdf".to_string(), Some("http://vtt".to_string()), "1:24:30".to_string(), 5)).unwrap(),
            "{\"Play\":[\"http://asdf\",\"http://vtt\",\"1:24:30\",5]}"
        );
        assert_eq!(
            serde_json::to_string(&HostMsg::Play("http://asdf".to_string(), None, "1:24:30".to_string(), 5)).unwrap(),
            "{\"Play\":[\"http://asdf\",null,\"1:24:30\",5]}"
        );
        assert_eq!(
            serde_json::to_string(&HostMsg::Pause).unwrap(),
            "\"Pause\""
        );
    }
    #[test]
    fn joinmsg_rep() {
        assert_eq!(
            serde_json::to_string(&JoinerMsg::Play("http://asdf".to_string(), Some("http://vtt".to_string()), 120, 5000)).unwrap(),
            "{\"Play\":[\"http://asdf\",\"http://vtt\",120,5000]}"
        );
        assert_eq!(
            serde_json::to_string(&JoinerMsg::Play("http://asdf".to_string(), None, 120, 5000)).unwrap(),
            "{\"Play\":[\"http://asdf\",null,120,5000]}"
        );
        assert_eq!(
            serde_json::to_string(&JoinerMsg::Pause).unwrap(),
            "\"Pause\""
        );
    }
}


clients/join/index.js
var host = window.location.host;
const urlParams = new URLSearchParams(window.location.search);
const socket = new WebSocket(`wss://${host}/msg`);
vid = document.getElementById("thevideo");
var init = 0;

mkstatus = function (msg) {
    document.getElementById("status").innerHTML = msg;
};

vid.onplay = function() {
    if (init == 0) {
        init = 1;
        socket.send('ready');
        mkstatus("Ready!");
        // document.getElementById("status").childNodes[0].innerHTML = "Ready!";
    }
};


socket.addEventListener('open', function (event) {
    var joinReq = { Join: urlParams.get('key') };
    socket.send(JSON.stringify(joinReq));
});
socket.addEventListener('message', function (event) {
    console.log(event.data);
    cmd = JSON.parse(event.data);

    if ("Pause" == cmd) {
        vid.pause();
        mkstatus("Stream is paused.");
    } else if ("Play" in cmd) {
        var cmd_src = cmd.Play[0];
        var cmd_vtt = cmd.Play[1];
        var cmd_seek = cmd.Play[2];
        var cmd_time = cmd.Play[3];

        vid.pause();
        if (vid.src != cmd_src) {
            vid.src = cmd_src;
        }
        if (null != cmd_vtt) {
            if (vid.childElementCount == 1) {
                var track = vid.children[0];
                if (track.src != cmd_vtt) {
                    track.src = cmd_vtt;
                }
            } else {
                var track = document.createElement("track");
                track.kind = "captions";
                track.label = "On";
                track.src = cmd_vtt;
                vid.append(track);
            }
        } else {
            if (vid.childElementCount == 1) {
                var track = vid.children[0];
                track.remove();
            }
        }
        vid.currentTime = cmd_seek;

        var diff = new Date(cmd_time) - Date.now();
        var timeout = setTimeout(function(){
            vid.play();
            mkstatus("Stream is playing.");
        }, diff);
        document.getElementById("status").innerHTML =
            "Sync start in "
            + Math.ceil(diff / 1000)
            + " seconds.";
    } else {
        document.getElementById("status").innerHTML =
            "Unknown command: " + JSON.stringify(cmd);
    }
});
socket.addEventListener('close', (event) => {
    document.getElementById("status").innerHTML =
        "⚠️ Connection closed.  Refresh page?"
});

setInterval(function() { socket.send("ping") }, 5000 );
